src/gas_price.rs

```diff
@@ -0,0 +1,116 @@
+    pub eip1559: Option<GasPrice1559>,
+}
+
+impl GasPrice {
+    // Estimate the gas price based on the current network conditions (base_fee_per_gas)
+    // Beware that gas price for mined transaction could be different (because base_fee_per_gas 
+    // can change between estimation and mining the tx).
+    pub fn estimate_gas_price(&self) -> f64 {
+        if let Some(gas_price) = &self.eip1559 {
+            match gas_price
+                .max_fee_per_gas
+                .partial_cmp(&(gas_price.max_priority_fee_per_gas + gas_price.base_fee_per_gas))
+            {
+                Some(ordering) => match ordering {
+                    std::cmp::Ordering::Less | std::cmp::Ordering::Equal => {
+                        gas_price.max_fee_per_gas
+                    }
+                    std::cmp::Ordering::Greater => {
+                        gas_price.max_priority_fee_per_gas + gas_price.base_fee_per_gas
+                    }
+                },
+                None => gas_price.max_fee_per_gas,
+            }
+        } else {
+            self.legacy
+        }
+    }
+
+    // Maximum gas price willing to pay for the transaction.
+    pub fn cap(&self) -> f64 {
+        if let Some(gas_price) = &self.eip1559 {
+            gas_price.max_fee_per_gas
+        } else {
+            self.legacy
+        }
+    }
+
+    // Bump maximum gas price by factor.
+    pub fn bump_cap(self, factor: f64) -> Self {
+        Self {
+            legacy: self.legacy * factor,
+            eip1559: self.eip1559.and_then(|x| Some(x.bump_cap(factor))),
+        }
+    }
+
+    // Ceil maximum gas price (since its defined as float).
+    pub fn ceil_cap(self) -> Self {
+        Self {
+            legacy: self.legacy.ceil(),
+            eip1559: self.eip1559.and_then(|x| Some(x.ceil_cap())),
+        }
+    }
+
+    // If current cap if higher then the input, set to input.
+    pub fn limit_cap(self, cap: f64) -> Self {
+        Self {
+            legacy: self.legacy.min(cap),
+            eip1559: self.eip1559.and_then(|x| Some(x.limit_cap(cap))),
+        }
+    }
+}
+
+/// Gas price structure for 1559 transactions. 
+/// Contains base_fee_per_gas as an essential part of the gas price estimation.
+#[derive(Debug, Default, Clone, Copy, PartialEq, PartialOrd)]
+pub struct GasPrice1559 {
+    // Estimated base fee for the pending block (block currently being mined)
+    pub base_fee_per_gas: f64,
+    // Maximum gas price willing to pay for the transaction.
+    pub max_fee_per_gas: f64,
+    // Priority fee used to incentivize miners to include the tx in case of network congestion.
+    pub max_priority_fee_per_gas: f64,
+}
+
+impl GasPrice1559 {
+    // Bump maximum gas price by factor.
+    pub fn bump_cap(self, factor: f64) -> Self {
+        Self {
+            base_fee_per_gas: self.base_fee_per_gas,
+            max_fee_per_gas: self.max_fee_per_gas * factor,
+            max_priority_fee_per_gas: self.max_priority_fee_per_gas,
```

Comment:
ubernit: You could also write this as:
```suggestion
            max_fee_per_gas: self.max_fee_per_gas * factor,
            ..self
```

This way, if the other fields change names, or if a new field gets added, there is less overhead in the changes.

Reply:
True, fixed.